### PR TITLE
Add badgestate

### PR DESCRIPTION
**Pull request by mlsteele**
Adds `BadgeState` and keeps it up to date. For the badging story this covers `gregor -> service` but does _not_ do anything for `service -> electron`.

`BadgeState` is just data and could be used from gregor in the future if need be. `Badger` holds a `BadgeState` and ties in to the service to keep it up to date.

I tested this out locally by having BadgeState write to a file and watching it. That code is removed in the second commit.

The next step after this PR is to make `Badger` export the badge state as a new protocol record through something like `NotifyRouter` to electron whenever it changes. (Here https://github.com/keybase/client/pull/4876)

r? @mmaxim

### CODE REVIEW

File: go/service/gregor.go [hunk 1]
```diff
@@ -249,6 +251,13 @@ func (g *gregorHandler) Connect(uri *rpc.FMPURI) (err error) {
 	} else {
 		err = g.connectNoTLS(uri)
 	}
+
+	if g.badger != nil {
```
Comment (mmaxim) [inline go/service/gregor.go]: Shouldn't this happen anytime we reconnect?

**Comment by mlsteele** [comment]
Yes, it should happen on reconnect! Moved that resync to `onConnect` and tested restarting gregord, now they do resync.

Also rebased.

File: go/service/gregor.go [hunk 2]
```diff
@@ -467,6 +461,12 @@ func (g *gregorHandler) OnConnect(ctx context.Context, conn *rpc.Connection,
 			len(replayedMsgs), len(consumedMsgs))
 	}
 
+	if g.badger != nil {
+		go func(badger *Badger) {
+			badger.Resync(context.TODO(), &chat1.RemoteClient{Cli: g.cli})
```
Comment (mmaxim) [inline go/service/gregor.go]: You should use the `Background` context here.

**Comment by mmaxim** [comment]
Is there anyway to test this stuff?

**Comment by mlsteele** [comment]
@mmaxim Added some (not great) tests.
Also moved this call to `pushState` https://github.com/keybase/client/pull/4873/commits/7023a75f0144024009d3afcaa14d213859cdf79a because the gregor tests don't hit `OnConnect`, only `serverSync` which is called by `OnConnect`. I think it makes sense to `pushState` after the sync anyway.

**Comment by mmaxim** [comment]
👍

**Comment by keybase-ci-visdiff** [comment]
The commits 705d92387c91...e7aaa143332a introduce visual changes on linux. <details><summary>:mag_right: 2 changed</summary>

 * Changed: **unlock_folders-Single** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479769428702-705d92387c91-e7aaa143332a-linux/unlock_folders-Single-705d92387c91.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479769428702-705d92387c91-e7aaa143332a-linux/unlock_folders-Single-e7aaa143332a.png)  
   unlock_folders-Single comparison
 * Changed: **unlock_folders-No_paperkeys** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479769428702-705d92387c91-e7aaa143332a-linux/unlock_folders-No_paperkeys-705d92387c91.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1479769428702-705d92387c91-e7aaa143332a-linux/unlock_folders-No_paperkeys-e7aaa143332a.png)  
   unlock_folders-No_paperkeys comparison
</summary>

**Pull request merged**